File: lowest-common-ancestor-of-a-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    private boolean getPath(TreeNode root, TreeNode p, List<TreeNode> path) {
        if (root == p) {
            return true;
        }
        if (root.left != null) {
            path.add(root.left);
            if (getPath(root.left, p, path)) return true;
            path.remove(path.size() - 1);
        }
        if (root.right != null) {
            path.add(root.right);
            if (getPath(root.right, p, path)) return true;
            path.remove(path.size() - 1);
        }
        return false;
    } 
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if (root == null || p == null || q == null) return null;
        List<TreeNode> pathP = new ArrayList<TreeNode>();
        List<TreeNode> pathQ = new ArrayList<TreeNode>();
        pathP.add(root);
        pathQ.add(root);
        if(!getPath(root, p, pathP)) return null;
        if(!getPath(root, q, pathQ)) return null;
        TreeNode lca = null;
        for (int i = 0; i < pathP.size() && i < pathQ.size(); ++i) {
            if (pathP.get(i).equals(pathQ.get(i))) lca = pathP.get(i);
        }
        return lca;
    }
    public TreeNode lowestCommonAncestor2(TreeNode root, TreeNode p, TreeNode q) {
        if (root == null) return null;
        if (root == p || root == q) return root;//p是q的儿子或者q是p的儿子会走到这个分支。
        TreeNode left = lowestCommonAncestor(root.left, p, q);
        TreeNode right = lowestCommonAncestor(root.right, p, q);
        if (left != null && right != null) return root; // 返回这个节点，两个p和q分别是它的左边和右边儿子。
        return left != null ? left : (right != null ? right : (null));
        /*返回不是null那边的结果，否则结果覆盖了。
        1. 返回的结果两边都为null，则表示root分支下没有解。
        2. 返回的结果有一个非空，则这个节点一定有正确地解。
        */
    }
}

File: src/main/cpp/lowest-common-ancestor-of-a-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
map<TreeNode*, TreeNode*>pa;
TreeNode* vec1[100005];int cnt1 = 0;
TreeNode* vec2[100005];int cnt2 = 0;
class Solution {
public:
    void dfs(TreeNode* root, TreeNode* pre) {
        if (root == NULL) return;
        pa[root] = pre;
        dfs(root -> left, root);
        dfs(root -> right, root);
    }
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        dfs(root, root);
        TreeNode* p1 = p;
        if (p == root) return root;
        if (q == root) return root;
        while (p1 != root) {
            vec1[cnt1++] = p1;
            p1 = pa[p1];
        }
        vec1[cnt1++] = p1;
        TreeNode* q1 = q;
        while (q1 != root) {
            vec2[cnt2++] = q1;
            q1 = pa[q1];
        }
        vec2[cnt2++] = q1;
        int i, j;
        for (i = cnt1 -1, j = cnt2 - 1; i >= 0 && j >= 0; i--, j--) {
            if (vec1[i] != vec2[j]) break;
        }
        return vec1[i+1];
    }
};